### DISCUSSION TITLE
devenv.sh support for OpenMage (next level development environment)

### DISCUSSION THREAD

**Discussion by fballiano** [OpenMage/magento-lts#2995] [Show & Tell]
In the past few days I've been working on a very cool (I think) project: https://github.com/fballiano/openmage-devenv

[devenv](https://devenv.sh) it's a powerful development environment based on [NixOS](https://nixos.org).  
It allows you to have containerized environments without containers or hypervisor or emulation, with native performance on any platform.  
This is by far the fasted Magento/Openmage development environment I've ever worked with and it's more than worth of the time to learn it.

I came across devenv just a few days ago and I've been blown away by it, and other big projects (like shopware) are already migrating to it.

I'd like to integrate it directly in OM's main repo but I'd wait a bit for that since it's fairly new.

If you want to know more about NixOS/devenv or this small project please let me know, I'm already migrating all of the dev environments and I think you'll have fun with it too.
Fabrizio

**Comment by justinbeaty** [comment]
Interesting, I'll try this out tomorrow. I am so behind devops tech, barely using docker, and now there's devenv. Last I used something like this was when I was using Vagrant to develop mage sites.

**Reply by fballiano** [reply]
not to have to use containers/hypervisors etc is game changing to me

**Reply by justinbeaty** [reply]
This is going to be a rabbit hole for me, lol. It's tricky because for anything other than vanilla OM development, I'm actually running OM headless with a VueJS frontend (and nodejs api layer).

**Reply by justinbeaty** [reply]
When running `devenv up` on Ubuntu 20.04, I get:

```
$ devenv up
Building shell ...
/nix/store/h8gvq6r4hgpa71h44dmg9qfx03mj81sv-coreutils-9.1/bin/readlink: error while loading shared libraries: libstdc++.so.6: cannot open shared object file: No such file or directory
/nix/store/wim4mqpn8lxhhr10p2kd070hyj152lil-bash-5.1-p16/bin/bash: error while loading shared libraries: libstdc++.so.6: cannot open shared object file: No such file or directory
```

I've tried some things to fix it. Not sure if it's a local problem or not yet.

**Reply by fballiano** [reply]
does "devenv shell" work my any means? it seems an error with nix itself

**Reply by justinbeaty** [reply]
No, it also did not work. I tried:

- Switch from `nixos-unstable` to `nixos-22.11`
- `sudo apt install libstdc++6` and `sudo apt install g++-multilib`
- Adding `buildInputs = [ stdenv.cc.cc.lib ];` in various places -- I'm not sure where that goes.

**Reply by domenkozar** [reply]
@justinbeaty could you share the output of:

- `$ ldd /nix/store/h8gvq6r4hgpa71h44dmg9qfx03mj81sv-coreutils-9.1/bin/readlink`
- `$ ldd /nix/store/wim4mqpn8lxhhr10p2kd070hyj152lil-bash-5.1-p16/bin/bash`

**Reply by justinbeaty** [reply]
@domenkozar I also should mention this is a virtualbox Ubuntu 20.04 guest on a Windows 10 host, in case it's relevant.
```
justin@dev:~$ ldd /nix/store/h8gvq6r4hgpa71h44dmg9qfx03mj81sv-coreutils-9.1/bin/readlink
	linux-vdso.so.1 (0x00007ffe59c5e000)
	/usr/lib/x86_64-linux-gnu/libjemalloc.so.2 (0x00007fefafb4c000)
	libacl.so.1 => /nix/store/dnwi357xww3rxdvr79gbvvs8zqm1x47m-acl-2.3.1/lib/libacl.so.1 (0x00007fefafb41000)
	libattr.so.1 => /nix/store/3nxvwpxmvv5sv2bb88plhs4ys1jl0bmm-attr-2.5.1/lib/libattr.so.1 (0x00007fefafb39000)
	libgmp.so.10 => /nix/store/j4mx37iy9lkkvcg41c9347sn7zwvnnw4-gmp-with-cxx-stage4-6.2.1/lib/libgmp.so.10 (0x00007fefafa99000)
	libc.so.6 => /nix/store/ayfr5l52xkqqjn3n4h9jfacgnchz1z7s-glibc-2.35-224/lib/libc.so.6 (0x00007fefaf890000)
	libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007fefaf728000)
	libstdc++.so.6 => /lib/x86_64-linux-gnu/libstdc++.so.6 (0x00007fefaf546000)
	libdl.so.2 => /lib/x86_64-linux-gnu/libdl.so.2 (0x00007fefaf540000)
	libgcc_s.so.1 => /lib/x86_64-linux-gnu/libgcc_s.so.1 (0x00007fefaf525000)
	libpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0 (0x00007fefaf502000)
	/nix/store/ayfr5l52xkqqjn3n4h9jfacgnchz1z7s-glibc-2.35-224/lib/ld-linux-x86-64.so.2 => /lib64/ld-linux-x86-64.so.2 (0x00007fefafe2a000)
justin@dev:~$ ldd /nix/store/wim4mqpn8lxhhr10p2kd070hyj152lil-bash-5.1-p16/bin/bash
	linux-vdso.so.1 (0x00007ffeb129b000)
	/usr/lib/x86_64-linux-gnu/libjemalloc.so.2 (0x00007f4f225f9000)
	libdl.so.2 => /nix/store/9xfad3b5z4y00mzmk2wnn4900q0qmxns-glibc-2.35-224/lib/libdl.so.2 (0x00007f4f225f4000)
	libc.so.6 => /nix/store/9xfad3b5z4y00mzmk2wnn4900q0qmxns-glibc-2.35-224/lib/libc.so.6 (0x00007f4f223eb000)
	libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007f4f22285000)
	libstdc++.so.6 => /lib/x86_64-linux-gnu/libstdc++.so.6 (0x00007f4f220a3000)
	libgcc_s.so.1 => /lib/x86_64-linux-gnu/libgcc_s.so.1 (0x00007f4f22088000)
	libpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0 (0x00007f4f22063000)
	/nix/store/9xfad3b5z4y00mzmk2wnn4900q0qmxns-glibc-2.35-224/lib/ld-linux-x86-64.so.2 => /lib64/ld-linux-x86-64.so.2 (0x00007f4f228d7000)
```

**Reply by Mic92** [reply]
the ldd you used is not the glibc ldd from nixpkgs. That's why it may print wrong results especially for glibc's own libraries. Run the following instead:

```
nix-shell -p glibc --run 'ldd /nix/store/h8gvq6r4hgpa71h44dmg9qfx03mj81sv-coreutils-9.1/bin/readlink'
```

Also you can generally get better debugging information in your case by doing `LD_DEBUG=libs devenv up` as this will print debug information from the link-loading process.

**Reply by justinbeaty** [reply]
It seems the culprit was that I had a custom `export LD_LIBRARY_PATH` in my `.bashrc`. I commented that out, and I was able to run `devenv up`.  `LD_DEBUG=libs devenv up` did help to pinpoint this.

**Reply by fballiano** [reply]
the guys at NIX/devenv are also extremely helpful

**Reply by justinbeaty** [reply]
@fballiano One other thing that maybe should go into your README; I had to change the ports so they don't conflict:

```
services.caddy.virtualHosts."http://openmage.test:5210" = {
services.redis.port = 5211;
services.mysql.settings.mysqld.port = 5212;
```

Or maybe the second two lines should be in your `devenv.nix`, but with `6379` and `3306` (the defaults)? Just so users don't have to google to see how to change the ports.

**Reply by fballiano** [reply]
just added those lines to my devenv.nix file ;-)

**Reply by justinbeaty** [reply]
I gotta learn more about nixos, seems like it could be a good tool.

**Comment by addison74** [comment]
As far as I am concerned, until last year in July I had OM installed in Ubuntu and with the help of PHPStrom I worked on a project directly in Linux. When Sven brought up DDEV, I started testing it intensively in Windows in various configurations. I found that installed in WSL2 is a little faster than when running in Ubuntu. I stayed in Windows since then, I installed PHPStorm along with other Windows-specific programs. DDEV is absolutely fantastic and configurable, but it requires resources, especially memory. Windows + Docker + PHPStorm sometimes reach 12 GB.

This solution proposed by you should also be tried so that we have as many installation and fast running options as possible for OM. A comparison with DDEV would be important, especially on the consumption of resources, but also on the custom side. From what I can see, the idea is similar to DDEV, where Docker is replaced by something else.

Maybe you should add phpMyAdmin in addition to Mailhog. DDEV has it and it is run with the command ```ddev launch -p```. DDEV allows installing Node.js for working with all stuff related to SASS/CSS/JS. There is also a useful extension called browsersync.

**Reply by fballiano** [reply]
docker is replaces by NixOS, which is just a package manager: https://nixos.org/guides/how-nix-works.html everything is native.

As far as I can see there's no phpmyadmin native option (I would have to install it which I'd prefer to avoid) but there's https://www.adminer.org although I much rather prefer to use a normal SQL client like https://sequel-ace.com/ on mac instead of a php solution, it's much more powerful.

I didn't include any node stuff to keep things slim but it's totally possible, I'll see what I can do about that.

sorry, a bit in a hurry

**Comment by fballiano** [comment]
My PR on NixOS packages to include n98-magerun and n98-magerun2 has been merged. I've included n98-magerun in my devenv environment, it's just great!
